pyomo/core/base/matrix_constraint.py

```diff
@@ -0,0 +1,360 @@
+        # fly do not have to here
+        body = self(exception=False)
+        if body is None:
+            return None
+        return min(self.lslack(), self.uslack())
+
+    #
+    # Override some default implementations on ComponentData
+    #
+
+    def index(self):
+        return self._index
+
+    #
+    # Abstract Interface (_ConstraintData)
+    #
+
+    @property
+    def body(self):
+        """Access the body of a constraint expression."""
+        comp = self.parent_component()
+        index = self._index
+        data = comp._A_data
+        indices = comp._A_indices
+        indptr = comp._A_indptr
+        x = comp._x
+        ptrs = xrange(indptr[index],
+                      indptr[index+1])
+        return LinearExpression(
+            linear_vars=[x[indices[p]] for p in ptrs],
+            linear_coefs=[data[p] for p in ptrs],
+            constant=0
+        )
+
+    @property
+    def lower(self):
+        """Access the lower bound of a constraint
+        expression."""
+        comp = self.parent_component()
+        index = self._index
+        return comp._lower[index]
+
+    @property
+    def upper(self):
+        """Access the upper bound of a constraint
+        expression."""
+        comp = self.parent_component()
+        index = self._index
+        return comp._upper[index]
+
+    @property
+    def equality(self):
+        """A boolean indicating whether this is an equality
+        constraint."""
+        comp = self.parent_component()
+        index = self._index
+        if (comp._lower[index] is None) or \
+           (comp._upper[index] is None):
+            return False
+        return comp._lower[index] == comp._upper[index]
+
+    @property
+    def strict_lower(self):
+        """A boolean indicating whether this constraint has
+        a strict lower bound."""
+        return False
+
+    @property
+    def strict_upper(self):
+        """A boolean indicating whether this constraint has
+        a strict upper bound."""
+        return False
+
+    def set_value(self, expr):
+        """Set the expression on this constraint."""
+        raise NotImplementedError(
+            "MatrixConstraint row elements can not be updated"
+        )
+
+@ModelComponentFactory.register(
+                   "A set of constraint expressions in Ax=b form.")
+class MatrixConstraint(collections_Mapping,
+                       IndexedConstraint):
+    """
+    Parameters
+    ----------
+    A_data : list
+        The values of the CSR format sparse matrix
+    A_indices : list
+        The column indices of the CSR format sparse matrix
+    A_indptr : list
+        The row start-stop pointers of the CSR format sparse matrix
+    lb : list
+        The list of constraint lower bounds
+    ub : list
+        The list of constraint upper bounds
+    x : list
+        The list of pyomo variables mapped to their appropriate column
+    """
+
+    def __init__(self, A_data, A_indices, A_indptr, lb, ub, x):
```

Comment:
I am a little concerned about the constructor API: For (most) Pyomo components, positional arguments are reserved for indexing sets and all other arguments must be passed by keyword argument.  This creates another exception to that rule (along with Piecewise, and to a certain extent, DerivativeVar)